### ISSUE TITLE
8th place solution [0.49 on private LB]

### ISSUE THREAD

**Issue by baekspace88** [KGGS/Open-Images-2019-Object-Detection#5]
https://www.kaggle.com/c/google-ai-open-images-object-detection-track/discussion/65120

**Comment by baekspace88** [comment]
- http://incredible.ai/deep-learning/2018/03/17/Faster-R-CNN/
- https://zzsza.github.io/data/2018/05/09/Faster-RCNN-review/
- https://yamalab.tistory.com/113
- https://towardsdatascience.com/faster-r-cnn-for-object-detection-a-technical-summary-474c5b857b46
- https://towardsdatascience.com/faster-r-cnn-object-detection-implemented-by-keras-for-custom-data-from-googles-open-images-125f62b9141a
-

**Comment by baekspace88** [comment]
## R-CNN
<img width="1009" alt="스크린샷 2020-02-14 오후 10 26 15" src="https://user-images.githubusercontent.com/48514061/74535415-1d557b80-4f79-11ea-8797-518c5b4a04b5.png">

- Regions with Convolutional Neural Networks
1. Image Input
2. Extract region proposals
  - Region proposal algorithm
  - 대상의 class를 구분하지 않고 이미지로부터 region을 추출해내는 과정
  - Selective search : 비슷한 질감이나 색, 강도를 가진 인접한 픽셀들을 연결하여 바운딩 박스를 구성하는 방법
  - 생성한 Bounding Boxes(각 이미지 당 2000개의 독립적인 Rectangular Boxes)를 압축(warp)하여 CNN 레이어의 Input으로 넣음.
3. Compute CNN features
  - Feature extraction from Region Proposals
  - Pre-trained 된 CNN 모듈에 통과시켜 새로운 features를 생성
  - Supervised Pre-training: ILSVRC2012 classification dataset으로 미리 학습된 AlexNet
  - Domain-Specific Fine-Tuning
    - 압축된 2000 region proposals를 실제 box(ground-truth instance)와의 IoU를 최대화 하는 방향으로 학습
  - 4096 dimensional feature vector를 최종적으로 얻음.

<img width="705" alt="스크린샷 2020-02-14 오후 11 03 16" src="https://user-images.githubusercontent.com/48514061/74537847-36acf680-4f7e-11ea-890c-9a4262712577.png">

4. Classify regions
  - SVM for object classification
    - 특정 클래스에 속한 object의 존재 여부를 검출
    - Feature vectors와 SVM weight의 스칼라 곱으로 분류
    - 이 단계를 나눠 놓은 이유는 도메인 목적에 맞는 분류기를 설정하고, 유효한 영역인지에 대해서만 판별하기 위함.
  - Bounding box regression
    - 결과물을 더 완벽하게 하기 위해, region proposal 단계에서 더욱 타이트한 bounding box를 만드는 단계
    - 필수적인 단계는 아니지만 만약 수행한다면 성능을 꽤 높여줄 수 있는 방법
 
<img width="796" alt="스크린샷 2020-02-14 오후 11 13 18" src="https://user-images.githubusercontent.com/48514061/74538452-9e177600-4f7f-11ea-96be-d57b9a06b8f7.png">

- https://towardsdatascience.com/r-cnn-for-object-detection-a-technical-summary-9e7bfa8a557c
- https://yamalab.tistory.com/110

**Comment by baekspace88** [comment]
## Fast R-CNN
<img width="635" alt="스크린샷 2020-02-15 오전 12 39 53" src="https://user-images.githubusercontent.com/48514061/74545148-b42b3380-4f8b-11ea-96dc-c72704b086a1.png">

- R-CNN의 단점 중 하나인 ‘속도’를 대폭 개선한 모델
- 입력 이미지에 한번만 CNN을 적용하고 RoI(Region of Interest) pooling으로 객체를 판별하기 위한 특징을 추출하는 구조
- ROI pooling layer : A special case of the spatial pyramid pooling (SPP) layer with just one pyramid level
   -  Bag-of-words (BoW) : 다양한 크기의 입력으로부터 일정한 크기의 feature를 추출해 낼 수 있는 방법 중 하나. 이미지가 지닌 특징들의 위치 정보를 모두 잃어버린다는 단점이 존재.
   -  Spatial Pyramid Pooling : 이미지를 여러개의 일정 개수의 지역으로 나눈 뒤, 각 지역에 BoW를 적용하여 지역적인 정보를 어느정도 유지할 수 있음.
   - Feature map 상의 특정 영역에 대해 일정한 고정된 개수의 bin으로 영역을 나눈 뒤, 각 bin에 대해 max pooling 또는 average pooling을 취함으로써 고정된 길이의 Feature vector를 얻을 수 있음.
<img width="550" alt="스크린샷 2020-02-15 오전 12 51 07" src="https://user-images.githubusercontent.com/48514061/74546046-454eda00-4f8d-11ea-9bd2-d361f7cca06c.png">
<img width="343" alt="스크린샷 2020-02-15 오전 12 52 56" src="https://user-images.githubusercontent.com/48514061/74546199-8a730c00-4f8d-11ea-8ce3-590735110ec4.png">

- ROI pooling layer의 output은 두 개의 FC layers를 거쳐 Softmax branch와 BB-regression branch에 각각 이력
- Softmax classification branch : 각 ROI가 K category와 하나의 catch-all background category에 속할 확률을 구함.
- BB regression branch : region proposal algorithm으로부터 얻은 bounding boxes를 더 정교하게 함.

- https://towardsdatascience.com/fast-r-cnn-for-object-detection-a-technical-summary-a0ff94faa022
- https://blog.lunit.io/2017/06/01/r-cnns-tutorial/

**Comment by baekspace88** [comment]
<img width="389" alt="KakaoTalk_Photo_2020-02-15-09-36-57" src="https://user-images.githubusercontent.com/48514061/74578234-c2527180-4fd6-11ea-8060-fd8a4aa1938b.png">

**Comment by baekspace88** [comment]
## Faster R-CNN
<img width="757" alt="스크린샷 2020-02-21 오후 4 48 46" src="https://user-images.githubusercontent.com/48514061/75014652-2343e300-54ca-11ea-8448-dc28f81712a8.png">

### Feature Extraction
- Base network(pre-trained Model)를 사용해 feature map 생성(해당 object를 학습해 놓은 상태)
- Input : 이미지, Output : feature map

### RPN(Region Proposal Network)
- Deep Convolution Network 구조
- Anchor : feature maps위에 n x n spatial window(보통 3 x 3)를 슬라이드하며 각 window 마다 한번에 여러개의 region proposals를 예측.
  - 보통 각 sliding window마다 9(k)개의 anchors가 존재하며, 3개의 서로 다른 종횡비 (aspect ratios) 그리고 3개의 서로 다른 크기(scales) 가 조합
<img width="610" alt="스크린샷 2020-02-21 오후 5 26 13" src="https://user-images.githubusercontent.com/48514061/75016872-47ee8980-54cf-11ea-8b13-1b28b5dcb441.png">

- Input : 이전 base network에서 뽑아낸 feature map, Output : 2개의 convolutional layers
  - Classification layer : anchor당 2개의 predictions값을 내놓으며, 객체인지 아니면 객체가 아닌지(그냥 배경인지)에 관한 확률값
    -  anchors 와 ground-truth boxes의 IoU를 통해 해당 anchors가 background인지 foreground인지 분류
  - Regression layer : 각 anchor당 델타 값들(x, y, width, height) (anchors에 적용이 되어서 최종 proposals)
    - anchors 와 ground-truth boxes의 각 벡터값을 이용해 학습
<img width="762" alt="스크린샷 2020-02-21 오후 5 28 33" src="https://user-images.githubusercontent.com/48514061/75017062-a6b40300-54cf-11ea-8a1b-035536f0ec8b.png">

### RoI(Region of Interest) Pooling
- 서로 다른 크기의 feature maps을 동일한 크기로 변환
  - 각각의 region proposal을 동일한 크기의 sections으로 나눔.(section의 크기는 RoI pooling의 output크기와 동일)
  - 각각의 section마다 가장 큰 값을 찾아 Output을 생성

### Object 감지 위한 Detector
- RoI풀링을 통과한 Output을 가지고 Object를 찾음

#### 참고 URL
- http://incredible.ai/deep-learning/2018/03/17/Faster-R-CNN/
- https://towardsdatascience.com/faster-r-cnn-for-object-detection-a-technical-summary-474c5b857b46
